include/deal.II/multigrid/mg_data_out.h

```diff
@@ -0,0 +1,153 @@
+
+#include <deal.II/numerics/data_out.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+/*!@addtogroup mg */
+/*@{*/
+
+/**
+ * Specialization of DataOut for output of all cells on a single level.
+ *
+ * This class allows for data output of all cells (active or not) on a given
+ * level. If the underlying Triangulation is parallel, only cells with the local
+ * level_subdomain_id are visualized.
+ *
+ * This class is particularly useful to output solutions on a
+ * specific multigrid level.
+ */
+template <int dim, typename DoFHandlerType = DoFHandler<dim>>
+class SingleLevelDataOut : public DataOut<dim, DoFHandlerType>
+{
+public:
+  /**
+   * Default constructor. Use reinit() to initialize the class.
+   */
+  SingleLevelDataOut()
+    : level(numbers::invalid_unsigned_int)
+  {}
+
+  /**
+   * Constructor.
+   * @param level Cells have to be on this level to be output.
+   */
+  explicit SingleLevelDataOut(const unsigned int level)
+    : level(level)
+  {}
+
+  /**
+   * Reset the options of this class (see the constructor above).
+   */
+  void
+  reinit(const unsigned int level)
+  {
+    this->level = level;
+  }
+
+  /**
+   * Implementation to get the first cell to output.
+   */
+  virtual typename DataOut<dim>::cell_iterator
+  first_cell() override
+  {
+    typename DataOut<dim>::cell_iterator cell =
+      this->triangulation->begin(level);
+    while ((cell != this->triangulation->end(level)) && !is_interesting(cell))
+      ++cell;
+
+    if (cell == this->triangulation->end(level))
+      return this->triangulation->end();
+
+    return cell;
+  }
+
+  /**
+   * Implementation to get the next cell to output.
+   */
+  virtual typename DataOut<dim>::cell_iterator
+  next_cell(const typename DataOut<dim>::cell_iterator &old_cell) override
+  {
+    if (old_cell != this->triangulation->end(level))
+      {
+        typename DataOut<dim>::cell_iterator cell = old_cell;
+        ++cell;
+        while ((cell != this->triangulation->end(level)) &&
+               !is_interesting(cell))
+          ++cell;
+        if (cell == this->triangulation->end(level))
+          return this->triangulation->end();
+
+        return cell;
+      }
+    else
+      return this->triangulation->end();
+  }
+
+
+  /**
+   * Implementation to get the first locally owned cell to output.
+   */
+  virtual typename DataOut<dim>::cell_iterator
+  first_locally_owned_cell() override
+  {
+    return first_cell();
+  }
+
+  /**
+   * Implementation to get the next locally owned cell to output.
+   */
+  virtual typename DataOut<dim>::cell_iterator
+  next_locally_owned_cell(
+    const typename DataOut<dim>::cell_iterator &old_cell) override
+  {
+    return next_cell(old_cell);
+  }
+
+private:
+  /**
+   * Internal function to check if this cell needs to be visualized.
+   */
+  bool
+  is_interesting(const typename DataOut<dim>::cell_iterator &cell) const
+  {
+    if (this->level != cell->level())
+      return false;
+
+    const types::subdomain_id level_subdomain_id =
```

Comment:
I got stuck on this function for a long time because I had assumed that `level_subdomain_id` is a property of the cell we're investigating. But that's not the case. Maybe rename it to `tria_subdomain_id` or `my_subdomain_id` or something like that?